cdap-client/src/main/java/co/cask/cdap/client/MetricsClient.java

```diff
@@ -50,21 +61,142 @@ public MetricsClient(ClientConfig config) {
   /**
    * Gets the value of a particular metric.
    *
-   * @param scope scope of the metric
    * @param context context of the metric
    * @param metric name of the metric
-   * @param timeRange time range to query
+   * @param groupBy how to group the metrics
    * @return value of the metric
    * @throws IOException if a network error occurred
    * @throws UnauthorizedException if the request is not authorized successfully in the gateway server
    */
-  // TODO: currently response from metrics endpoint is not "regular", so it's not easy to return an object from here
-  // (e.g. metrics endpoint sometimes returns {"data":0} and other times returns {"data":[..]})
-  public JsonObject getMetric(String scope, String context, String metric, String timeRange) throws IOException,
-    UnauthorizedException {
-    URL url = config.resolveURL(String.format("metrics/%s/%s/%s?%s", scope, context, metric, timeRange));
-    HttpResponse response = restClient.execute(HttpMethod.GET, url, config.getAccessToken());
-    return ObjectResponse.fromJsonBody(response, JsonObject.class).getResponseObject();
+  public MetricQueryResult query(Map<String, String> context, String metric, @Nullable String groupBy)
+    throws IOException, UnauthorizedException {
+
+    URL url = config.resolveURLV3(String.format("metrics/query?context=%s&metric=%s%s",
+                                                contextToPathParam(context), metric,
+                                                groupBy == null ? "" : "&groupBy=" + groupBy));
+    HttpResponse response = restClient.execute(HttpMethod.POST, url, config.getAccessToken());
+    return ObjectResponse.fromJsonBody(response, MetricQueryResult.class).getResponseObject();
+  }
+
```

Comment:
I'd prefer to add that in another PR since it isn't necessary for https://github.com/caskdata/cdap/pull/1567